### ISSUE TITLE
Video download problem - "Unauthorized Access","code":101

### ISSUE THREAD

**Issue by yzhang3000** [fronzbot/blinkpy#422]
I have been using Blinkpy to download videos to my PC for almost a year. Last time that I was able to download the video files was about 4 weeks ago. Today when I was trying to download new video files since my last downloading, I found the sizes of the downloaded video files are all 1kB, obviously something was not right. When I open the video files they just have one line of ASCII text showing below:
{"message":"Unauthorized Access","code":101}
My user name and password should be correct because after the code:
>blink.auth = auth
>blink.start()
No error message is shown. If the password is not correct, the error message 'Login endpoint failed. Try again later.' will be shown. 

Anybody know what is the csuse of this problem? Thank you.

**Comment by fronzbot** [comment]
A few things could be happening, here are my too two guesses:

1. Your account was blocked by Blink servers for excessive API usage. You can verify this by opening the Blink app on your phone. If you can log in there, this is not the issue.

2. The video API endpoint changed and the one being used now no longer works. I should be able to verify this by checking if I get the same result on my end (just need to find some time to do testing)

Would you be willing to post the full code you used (obviously removing username, password, etc)? That way I (or someone else) can attempt to exactly replicate the behavior you're seeing?

**Comment by yzhang3000** [comment]
After I got this problem, I can still access my account using Blink app with iphone.  I can access all the new videos on the cloud using my phone and access the live steams of my 3 cameras using my phone.

My code is pretty simple, just like this:

import blinkpy
from blinkpy.blinkpy import Blink
from blinkpy.auth import Auth

blink = Blink()
auth = Auth({"username": "xxx@xxx.com", "password": "********"}, no_prompt=True)
blink.auth = auth
blink.start()

blink.download_videos(path='e:/blink', since='2021/01/01', debug=False)

**Comment by fronzbot** [comment]
Looks like you're ignoring the 2FA code, is that correct? Try setting `no_prompt=False` and entering the 2FA code that is emailed to you when prompted to see if that fixes the problem

**Comment by yzhang3000** [comment]
I was doing this for quite some time. I noticed that if I run the script to download large amount of videos (a few hundreds),  it will run for a while downloading some files and then stopped with some error messages, and I will receive multiple emails with "verification PIN" at the same time. Usually I just ignore those emails, wait for some time (like 10 minutes) and retry by changing the date range (beside the 'since', I added a 'until' variable to the blink.download_videos() so I can assign a date window for downloading).  I never understand how to use the "verification PIN" correctly with your code.  I tried something like this before:

#my_blink_key = '111111'
#auth.send_auth_key(blink, my_blink_key)
#blink.setup_post_verify()

And it has no effect.

**Comment by fronzbot** [comment]
Yeah that's how you do it. https://github.com/fronzbot/blinkpy#starting-blink-without-a-prompt

By saying "it had no effect" do you mean it didn't change the fact the outcome of downloading? It's possible you're simply being throttled. Downloading 100s of files might be a red flag to blink.

It might be useful to add a parameter to the blink download handler to set a small delay between video retrieval to prevent this :thinking:

**Comment by fronzbot** [comment]
There's actually another possibility here: the location of the video file is different than what is grabbed during the API query. What type of camera are you using? XT/XT2/Mini? Do you have multiple cameras? Are all failing or just certain ones?

**Comment by yzhang3000** [comment]
Yes, when I said "no effect" I mean it does not change the fact the outcome of downloading. 

Now I changed the 'no_prompt' to 'False', when I run

blink = Blink()
auth = Auth({"username": "xxx@xxx.com", "password": "********"}, no_prompt=False)
blink.auth = auth
blink.start()

The program will pause and prompting "Enter code sent to xxxx@xxx.com:",  and then I just type in the six digital pin code I received in my email. And program will keep running and finished the downloading. And the downloaded files are all 1kB size with the same content "...Unauthorized Access..."

**Comment by yzhang3000** [comment]
Now I tried both "Starting blink without a prompt + blink.auth.send_auth_key()" and "Starting blink with prompt + manually input key", and got the same results. So the problem looks like it's not caused by 2FA but something else. I believe your comment about being throttled for my previous experience is correct, and that's not a major issue since I could always wait for a while and resume the downloading. However, this new issue is preventing me from downloading any video at all. 

Would you please find the root cause and provide a solution? I really like your code and it's a very convenient solution for downloading Blink videos to PC. I really want to keep using it in the future. Thank you.

**Comment by rohitsud** [comment]
I can confirm seeing the 401 response even after the 2FA code. The setup was working fine some days ago.

The app loads videos fine though.

**Comment by briantroy** [comment]
I can confirm. App works fine but all video files download are simple text files (json really) containing:
{"message":"Unauthorized Access","code":101}

This started on the 28th around 4am Pacific time:

-rw-rw-r-- 1 user.group 110K Jan 27 06:25 driveway-2021-01-27t14-20-49-00-00.mp4
-rw-rw-r-- 1 user.group  330K Jan 27 12:56 driveway-2021-01-27t20-54-59-00-00.mp4
-rw-rw-r-- 1 user.group  229K Jan 28 03:55 driveway-2021-01-28t11-50-48-00-00.mp4
-rw-rw-r-- 1 user.group  161K Jan 28 03:55 driveway-2021-01-28t11-51-09-00-00.mp4
-rw-rw-r-- 1 user.group   44 Jan 28 15:30 driveway-2021-01-28t23-28-42-00-00.mp4
-rw-rw-r-- 1 user.group    44 Jan 28 17:55 driveway-2021-01-29t01-52-47-00-00.mp4
-rw-rw-r-- 1 user.group    44 Jan 29 08:01 driveway-2021-01-29t16-00-51-00-00.mp4


I stopped the process 24 hours ago, manually refreshed credentials to the cache file - and then restarted. Same.
I stopped the process last night (12+ hours) then restarted. Same.

Still image capture works fine and did for the entire time between the 28th and yesterday when videos stopped working.

Has anyone confirmed they have downloaded videos after the 28th?

5 XT cameras which have been working with the process running for months.

**Comment by secops-testing** [comment]
I've been getting the "unauthorized access" results ever since the 27th.  The last actual video I managed to get was from the 26th.  I can still access from the app just fine, so I'm not blacklisted.

However, I can't download any videos at all.

**Comment by ossumpossum** [comment]
I've been getting the "unauthorized access" results ever since the 27th. The last actual video I managed to get was from the 26th. I can still access from the app just fine, so I'm not blacklisted.

However, I can't download any videos at all.

**Comment by alexmakdev** [comment]
I can confirm as well with not being able to download any videos.

**Comment by rbrtrs** [comment]
Hi, I can also confirm this behavior starting on Jan 29th. At the same time 2FA stopped working. Verify requests do fail using the correct PIN. I think this is also related to https://github.com/fronzbot/blinkpy/issues/418. They probably have changed the 2FA API.

**Comment by clearnano** [comment]
I can also confirm when running 
blink.download_videos('/home/larry/blink', since='2021/02/01', camera='all', stop=10, debug=False)
the video files and names download in a .mp4 extension but it's only 44bytes and not playable.

If we can look into this that'd be phenomenal

No error messages seen 
Blink XT and Blink outdoor cameras and sync module 2

**Comment by fronzbot** [comment]
Could be an API endpoint changed or the auth flow has been modified. Not sure on ETA of solution

**Comment by briantroy** [comment]
An interesting additional bit of information. Prior to this the videos would be marked as viewed in the app - that no longer happens either. So it is more than just “download” on the Blink side...

**Comment by macleajb** [comment]
If I use the curl raw way to get a token and then use that token to replace what blinkpy generates/saves by replacing it in the file I call  with : 

auth = Auth(json_load("config.json"))

Then the download call works.

Otherwise I too get the same error now.

**Comment by pferland** [comment]
@macleajb can you give a little bit of a bigger code snippet on how you are managing the auth token?

**Comment by macleajb** [comment]
@pferland : 
Make the auth request:
curl 'https://rest-prod.immedia-semi.com/api/v4/account/login' -X POST  -H 'Content-Type: application/json' -d '{"unique_id": "19283746-0000-0000-0000-000000000000", "password":"Your_Password","email":"your@email"}'
This return in JSON format, the info including the token and region you need to adjust your URL requests with. Appears to also include if you need to verify the account: 
"verification_required":true

From that I do the verify:
curl --request POST --url https://rest-{region}.immedia-semi.com/api/v4/account/{account id}/client/{client id}/pin/verify --header 'content-type: application/json' --header 'token-auth: {Token}' --data '{"pin":"{PIN sent to phone}"}'
You should get a positive response.

Then I just replace the "token": value in the json file I load for blink. You can also just confirm it is working with more requests: 

curl --request GET --url 'https://rest-{region}.immedia-semi.com/api/v1/accounts/{account id}/media/changed?since=2021-02-02T09%3A58%3A14%2B0000&page=1' --header 'token-auth: {token}'

I notice that once I have that token, I no longer need to do the auth call, but, if I do, it may send me a PIN even though the response says "verification_required":false.

**Comment by fronzbot** [comment]
@macleajb - those curl requests are precisely what the library is doing via the python requests library. Since we get different results between the two methods, there must be _something_ different between what you're doing manually and what the library is automating. Most of the API calls are implemented in the api.py module, so any extra set of eyes to find differences would be helpful!

**Comment by pferland** [comment]
@macleajb 
Weird, I get invalid pin when I try to verify it. Maybe I did it slightly wrong?

> [pferland@piecave blinkpy]$ curl -H "Host: rest.prod.immedia-semi.com" -H "Content-Type: application/json" --data-binary '{ "password" : "<PASSWD>", "client_specifier" : "Fedora 32", "email" : "<LOGIN>" }' --compressed https://rest.prod.immedia-semi.com/api/v4/account/login -k
> 
> {"account":{"id":<USER_ID>,"verification_required":false,"new_account":false},"client":{"id":<CLIENT_ID>,"verification_required":true},"authtoken":{"authtoken":"<AUTHTOKEN>","message":"auth"},"region":{"tier":"u026","description":"United States","code":"ap"},"lockout_time_remaining":0,"force_password_reset":false,"allow_pin_resend_seconds":60}
> 
> [pferland@piecave blinkpy]$ curl --request POST --url https://rest-u026.immedia-semi.com/api/v4/account/<USER_ID>/client/<CLIENT_ID>/pin/verify --header 'content-type: application/json' --header 'token-auth: <AUTHTOKEN>' --data '{"pin":"<EMAIL_PIN>"}'
> {"valid":false,"require_new_pin":false,"code":1621,"message":"Invalid PIN"}
> [pferland@piecave blinkpy]$

I have noticed though, that the PIN verification is not required and you can do other REST API calls as noted in my other issues ticket #418.

> curl -H "Host: rest.u026.immedia-semi.com" -H "TOKEN_AUTH: <AUTHTOKEN FROM ABOVE>" --compressed https://rest.u026.immedia-semi.com/networks -k
> {"summary":{"<NETWORK_ID>":{"name":"<NETWORKNAME>","onboarded":true},"<NETWORK_ID>":{"name":"<NETWORKNAME> ","onboarded":true}},"networks":[{"id":<NETWORK_ID>,"created_at":"2020-11-25T15:06:35+00:00","updated_at":"2020-11-25T15:06:35+00:00" <TRUNCATED>

**Comment by pferland** [comment]
I can list out the videos, but when doing an access to the video file itself, I get Unauthroized access.

[pferland@piecave blinkpy]$ curl --request GET --url 'https://rest-u026.immedia-semi.com/api/v2/accounts/<USER_ID>/media/clip/<FILE_NAME>.mp4' --header 'token-auth: <AUTH_TOKEN>' -k
{"message":"Unauthorized Access","code":101}


Same thing for the Tumbnails

[pferland@piecave blinkpy]$ curl --request GET --url 'https://rest-u026.immedia-semi.com/api/v2/accounts/<USER_ID>/media/thumb/<THUMB_NAME>' --header 'token-auth: <AUTH_TOKEN>' -k
{"message":"Unauthorized Access","code":101}

**Comment by rbrtrs** [comment]
I also still get {"valid":false} responses. Tried to modify the header a bit, set explicit POST request type etc., all using plain curl, not Python. Still verification fails. Listing media files works even without verification, but not downloading.

Would be happy to help, but don't know how do you determine the correct API endpoints in case the have been changed.

**Comment by rbrtrs** [comment]
@macleajb: Are you sure that the downloaded files are valid MP4 files?

**Comment by macleajb** [comment]
> @macleajb: Are you sure that the downloaded files are valid MP4 files?

Yes I can confirm the mp4's are proper. It appears if you don't get a successful pin verify you can still do some of the queries but the MP4 downloads will fail...

**Comment by pferland** [comment]
@macleajb Interesting, I wonder how you are getting the 2FA pin validated and I can't.

**Comment by macleajb** [comment]
@pferland, that sure looks correct to me :(. 
> > [pferland@piecave blinkpy]$ curl --request POST --url https://rest-u026.immedia-semi.com/api/v4/account/USER_ID/client/CLIENT_ID/pin/verify --header 'content-type: application/json' --header 'token-auth: AUTHTOKEN' --data '{"pin":"EMAIL_PIN"}'
> > {"valid":false,"require_new_pin":false,"code":1621,"message":"Invalid PIN"}

Hmm, but I use unique_id: when I login, not client_specifier. Not sure what difference that would make.

**Comment by pferland** [comment]
Can confirm a valid 2FA validation will allow the video file to be downloaded.


> [pferland@piecave blinkpy]$ curl -H "Host: rest.prod.immedia-semi.com" -H "Content-Type: application/json" --data-binary '{ "password" : "PASSWD", "client_specifier" : "Fedora 32", "email" : "LOGIN" }' --compressed https://rest.prod.immedia-semi.com/api/v4/account/login -k
> {"account":{"id":USER_ID,"verification_required":false,"new_account":false},"client":{"id":CLIENT_ID,"verification_required":true},"authtoken":{"authtoken":"AUTH_TOKEN","message":"auth"},"region":{"tier":"u026","description":"United States","code":"ap"},"lockout_time_remaining":0,"force_password_reset":false,"allow_pin_resend_seconds":60}
> 
> [pferland@piecave blinkpy]$ curl --request POST --url https://rest-u026.immedia-semi.com/api/v4/account/USER_ID/client/CLIENT_ID/pin/verify --header 'content-type: application/json' --header 'token-auth: AUTH_TOKEN' --data '{"pin":"EMAIL_PIN"}'
> {"valid":true,"require_new_pin":false,"message":"Client has been successfully verified","code":1626}
> 
> [pferland@piecave blinkpy]$ curl --request GET --url 'https://rest-u026.immedia-semi.com/api/v2/accounts/USER_ID/media/clip/FILE_NAME.mp4' --header 'token-auth: AUTH_TOKEN' -k --output /home/pferland/test.mp4
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100 7238k  100 7238k    0     0  17.1M      0 --:--:-- --:--:-- --:--:-- 17.1M

**Comment by macleajb** [comment]
> Enter code sent to <EMAIL>: 971337
> Did not receive valid response from server.

@pferland , not sure why it would ask for your code as the token you entered via the load() call would be valid and not need to do a login() call. Almost sounds like you dod not load the new token or it is not be presented properly.

**Comment by fronzbot** [comment]
Sounds like something in the library is broken regarding logging in/sending 2FA pin. Not necessarily and api endpoint change, but more of a payload discrepancy problem?

**Comment by pferland** [comment]
@macleajb 

My issue was I was saving my config and it was using an old config. Cleared out that config and re-ran, verify says it worked, but the videos are still not downloading properly, they are all 44 bytes.

**Comment by rbrtrs** [comment]
Now I don't even get verification e-mails with a PIN anymore. App still works fine.

**Comment by pferland** [comment]
@rbrtrs I noticed that when the session info is saved, a new PIN will not be sent.

delete the blinkpy session file (if you have one, this happened when I had an old session file saved.)
re-run the script, and a new 2FA Pin should be sent.

**Comment by rbrtrs** [comment]
@pferland: Thank you for the hint. But I am still using plain curl calls for testing, not blinkpy. Tested with two different accounts. Very strange.

**Comment by pferland** [comment]
@rbrtrs I do know they limit emails being sent out, maybe wait half hour to an hour or so and try it again?

**Comment by rbrtrs** [comment]
@pferland: Yes, will try at a later point in time again.

**Comment by pferland** [comment]
Interesting, I was messing around with pythons requests module to see if that would make things work, but its really odd that I get a 404 response code for the Video List URL, but the command line curl will still work...



```
import json
from pprint import pprint
import requests
data = {"password": "<PASSWD>", "client_specifier": "Fedora 32", "email": "<LOGIN>"}
response = requests.post('https://rest.prod.immedia-semi.com/api/v4/account/login', json=data, verify=False)
if response.status_code is 200:
    data = json.loads(response.text)
    pprint(data)

    auth_token = str(data['authtoken']['authtoken'])
    user_id = str(data['account']['id'])
    client_id = str(data['client']['id'])
    verification_required = data['client']['verification_required']
    region_server = str(data['region']['tier'])
    #headers = {'User-Agent': 'curl/7.54', 'token-auth': auth_token}
    headers = {'token-auth': auth_token}
    pprint(auth_token)

    if verification_required is True:
        print("Enter 2FA Pin from Blink Email: ")
        pin_input = input()

        data = {"pin": pin_input}
        response2 = requests.post('https://rest.'+region_server+'.immedia-semi.com/api/v4/account/'+user_id+'/client/'+client_id+'/pin/verify', json=data, verify=False, headers=headers)
        data2 = json.loads(response2.text)
        pprint(data2)
        if data2['valid'] is True:
            pprint('https://rest.' + region_server + '.immedia-semi.com/api/v1/accounts/'+user_id+'/media/changed?since=2020-12-02')
            response3 = requests.post(
                'https://rest.' + region_server + '.immedia-semi.com/api/v1/accounts/'+user_id+'/media/changed?since=2020-12-02&page=1',
                json=data, verify=False, headers=headers)
            pprint(response3.status_code)

        else:
            sys.exit("Failed 2FA Pin Verification")

    else:
        sys.exit("Verification not required?")

else:
    sys.exit("Failed to Login")

```

output: 


```

Enter 2FA Pin from Blink Email: 
128417
{'code': 1626,
 'message': 'Client has been successfully verified',
 'require_new_pin': False,
 'valid': True}
'https://rest.u026.immedia-semi.com/api/v1/accounts/<USER_ID>/media/changed?since=2020-12-02'
404
```


Even weirder is that I get a 403 when trying to access the video file itself.

```
{'code': 1626,
 'message': 'Client has been successfully verified',
 'require_new_pin': False,
 'valid': True}
'https://rest.u026.immedia-semi.com/api/v2/accounts/<USER_ID>/media/clip/<FILE_NAME>.mp4'
403
```


I wonder if doing a direct curl call from python would make a difference, maybe it might have to be a bash script to get it to work too. Going to try a few things.

**Comment by pferland** [comment]
Got a working work around using bash and python together.

Will do a little clean up either tonight or tomorrow and post it here if you would like

**Comment by rbrtrs** [comment]
This would be highly appreciated by me. Was not able to follow up yet because the missing PIN notification...

**Comment by pferland** [comment]
Are you still able to do things on the Blink Mobile app? If not, I hope they didn't suspend your account for abusing the API :(

If you can, that's really weird how you have not gotten a new PIN email.

**Comment by rbrtrs** [comment]
I don't think they have suspended my account since in that case they would have suspended two independent accounts at the same time. The App is still working fine with both accounts.

**Comment by pferland** [comment]
okay, good to know. will work on cleaning the script up tomorrow and post it here.

**Comment by macleajb** [comment]
In blinkpy/auth.py, replace:
return {"TOKEN_AUTH": self.token, "user-agent": DEFAULT_USER_AGENT}
with
return {"TOKEN_AUTH": self.token, "user-agent": DEFAULT_USER_AGENT, "content-type": "application/json"}
build and run. 

PIN validation call succeeds. At least it does for me.

Existing problem is that validation fails if the content-type: application/json header is not passed in the call.

This is not a correct fix. I only submit it to suggest what appears to be the error. I did not investigate to see what harm forcing that content-type on for everything does and it may do just that (cause harm).

Cheers,
DrJES

**Comment by pferland** [comment]
Okay, the script is all cleaned up. It probably has a lot of things that can be done differently and more efficiently, but it was a quick and dirty script to get what we needed done.

Requirements: Python3 (json, argparse, sys) and /bin/bash.

**_download.sh script (main script to run):_**


```
#!/bin/bash
# Created By: Phillip Ferland
# Email: perland@randomintervals.com
# Date: 2021-02-05
# Description: Work around for Blink REST API to download videos.
#              Seems like 2FA Pin validation is required to get a valid back now inorder to download the videos.
#              But it seems to only work with raw cURL for some reason.
#              Python modules (that I have tried; pycurl and requests) get a 404 or 403, even after 2FA PIN validation


USERNAME=$1
PASSWORD=$2
VIDEO_FROM_DATE=$3
SAVE_FOLDER=$4
SLEEP_BETWEEN_VIDEO=$5

if [ $# -lt 4 ]; then
  echo "Not enough arguments."
  echo ""
  echo "Usage:"
  echo "./download.sh <USERNAME> <PASSWD> <CHANGES_SINCE_DATE> <SAVE_FOLDER> <SLEEP_BETWEEN_VIDEO>"
  echo ""
  echo "Example:"
  echo './download.sh pferland@randomintervals.com PA$$W)4D 2020-12-02 /mnt/blink_videos'
  echo ""
  echo "<SLEEP_BETWEEN_VIDEO> is optional, default is 60 seconds to be nice to the Blink REST API and we dont get kicked off it."
  exit
fi

if [ "${5}" == "" ]; then
  SLEEP_BETWEEN_VIDEO=60
else
  echo "SLEEP_BETWEEN_VIDEO set to: ${SLEEP_BETWEEN_VIDEO} seconds"
fi

LOGIN=$(curl --silent -H "Host: rest.prod.immedia-semi.com" -H "Content-Type: application/json" \
--data-binary '{ "password" : "'${PASSWORD}'", "client_specifier" : "Fedora 32", "email" : "'${USERNAME}'" }' \
--compressed https://rest.prod.immedia-semi.com/api/v4/account/login -k)

#echo "${LOGIN}"

# Parse for Auth Token
IN=$(echo "${LOGIN}" | jq -r '.authtoken' | grep "authtoken")
arrIN=(${IN//: / })
token=${arrIN[1]}
token=$(echo "$token" | sed 's/,//g' | sed 's/"//g')
#echo "${token}"
HEADER="token-auth: ${token}"

# Parse for User ID
IN=$(echo "${LOGIN}" | jq -r '.account' | grep "id")
arrIN=(${IN//: / })
USER_ID=${arrIN[1]}
USER_ID=$(echo "${USER_ID}" | sed 's/,//g')
#echo "${USER_ID}"

# Parse for Client ID
IN=$(echo "${LOGIN}" | jq -r '.client' | grep "id")
arrIN=(${IN//: / })
CLIENT_ID=${arrIN[1]}
CLIENT_ID=$(echo "${CLIENT_ID}" | sed 's/,//g')
#echo "${CLIENT_ID}"

#Parse for Validation required
IN=$(echo "${LOGIN}" | jq -r '.client' | grep "verification_required")
arrIN=(${IN//: / })
VER_REQ=${arrIN[1]}
VER_REQ=$(echo "${VER_REQ}" | sed 's/,//g')
#echo "${VER_REQ}"

#Parse for REST API Region Server
IN=$(echo "${LOGIN}" | jq -r '.region' | grep "tier")
arrIN=(${IN//: / })
REGION=${arrIN[1]}
REGION=$(echo "${REGION}" | sed 's/,//g' | sed 's/"//g')
#echo "${REGION}"

#exit
if [ "${VER_REQ}" == "true" ]; then
  echo "Need to verify Client."
  sleep 2
  read -r -p "Blink 2FA Pin: " PIN
  # 2FA Verification
  #echo "https://rest-${REGION}.immedia-semi.com/api/v4/account/${USER_ID}/client/${CLIENT_ID}/pin/verify"
  VERIF=$(curl --silent --request POST --url "https://rest-u026.immedia-semi.com/api/v4/account/${USER_ID}/client/${CLIENT_ID}/pin/verify" --header 'content-type: application/json' --header "token-auth: ${token}" --data '{"pin":"'${PIN}'"}')

  VALID=$(echo "${VERIF}" | jq -r '.valid')
#  echo ${VALID}
else
  echo "Verification not required."
  VALID="true"
fi

#echo $VALID

if [ "$VALID" = "true" ]; then
    echo "User 2FA Pin Validated."
    #List Videos
    LIST_VIDEOS=$(curl --silent --request GET --url "https://rest-${REGION}.immedia-semi.com/api/v1/accounts/${USER_ID}/media/changed?since=${VIDEO_FROM_DATE}&page=1" --header 'content-type: application/json' --header "token-auth: ${token}" -k)

    echo "Using Python to process the Video List (it was easier and quicker than using bash...)"
    python3 processes_list.py -i "${LIST_VIDEOS}"

    while IFS="" read -r p || [ -n "$p" ]
    do
      #printf '%s\n' "$p"
      arrIN=(${p//::/ })
      FILENAME=${arrIN[1]}
      SAVEFILE=${arrIN[0]}

      #echo $SAVEFILE
      #echo $FILENAME
      #Download Videos
      echo "/home/pferland/${SAVEFILE}"
      curl --request GET --url "https://rest-${REGION}.immedia-semi.com/api/v2/accounts/${USER_ID}/media/clip/${FILENAME}" --header "token-auth: ${token}" -k --output "${SAVE_FOLDER}/${SAVEFILE}"
      echo ""
      echo "||-------------------------------------------------------------------||"
      sleep ${SLEEP_BETWEEN_VIDEO}
    done < file_list.txt
    exit
else
    echo "User 2FA Pin Failed Validation."
    exit
fi

```



**_processes_list.py  (run by the bash script to parse the video files list)_**

```
import json
import argparse
import sys


def get_options(args=sys.argv[1:]):
    parser = argparse.ArgumentParser(description="Parses command.")
    parser.add_argument("-i", "--input", help="Your input string.")
    arg_options = parser.parse_args(args)
    return arg_options


options = get_options()

json_data = json.loads(options.input)

with open("file_list.txt", "w+") as FH:
    for media in json_data['media']:
        if media['deleted'] == "true":
            continue
        file_id = str(media['media']).split("/")[-1].split(".")[0]
        name = str(media['device_name']).replace(" - ", "-").replace(" ", "-")
        created = str(media['created_at']).replace("+", "-").replace(":", "-")
        save_filename = name+"-"+created+".mp4"
        #print(save_filename+"::"+file_id+"\n")
        FH.write(save_filename+"::"+file_id+"\n")
    FH.close()

```

**Comment by pferland** [comment]
Just realized that I did not check for existing videos if they were already downloaded. So this script will re-download videos that already are on the local filesystem.

**Comment by pferland** [comment]
Fixed the temp scripts so that they look for the existing video files and also does not attempt to download the deleted files.

**Comment by fronzbot** [comment]
Hi all,

This should now be fixed with the `0.16.5rc0` release. Once I get some confirmation from you that this fixed the issue, I'll push the final 0.16.5 release. To test, simple run `pip install blinkpy==0.16.5rc0`

Let me know if you're seeing any undesired behavior as well 👍

**Comment by pferland** [comment]
@fronzbot just installed the 0.16.5rc0 and it looks like it is downloading the videos properly now.

Just needed to delete the old session file and have it re-created with a validated 2FA pin.

**Comment by rbrtrs** [comment]
Thank you @pferland, @macleajb and @fronzbot!

**Comment by kwkid** [comment]
Newbie here so please bear with me. I've been following the instructions above with a successful install of 0.16.5.rc0. The part i can't seem to resolve is 'Just needed to delete the old session file '...@pferland would you be able to give the path and filename for this session file. I can't seem to locate it?

much appreciated

**Comment by pferland** [comment]
@kwkid If you use `blink.save(blink_conf)` that saves a session config file after a login and 2FA validation. If you have one of these files saved before the `0.16.5rc0` update, you will need to delete it first. If you do not save the session data with `blink.save(blink_conf)` you do not have to do anything and can use `0.16.5rc0` without any pre-req steps.

**Comment by kwkid** [comment]
@pferland....Ah makes total sense.....cheers
